azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeServiceAdapter.java

```diff
@@ -0,0 +1,429 @@
+import org.jclouds.azurecompute.arm.domain.SKU;
+import org.jclouds.azurecompute.arm.domain.VMDeployment;
+import org.jclouds.azurecompute.arm.domain.VMSize;
+import org.jclouds.azurecompute.arm.features.OSImageApi;
+import org.jclouds.azurecompute.arm.features.ResourceGroupApi;
+import org.jclouds.azurecompute.arm.functions.ParseJobStatus;
+import org.jclouds.azurecompute.arm.util.DeploymentTemplateBuilder;
+import org.jclouds.compute.ComputeServiceAdapter;
+import org.jclouds.compute.domain.Template;
+import org.jclouds.compute.reference.ComputeServiceConstants;
+import org.jclouds.domain.LoginCredentials;
+import org.jclouds.logging.Logger;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import org.jclouds.util.Predicates2;
+
+/**
+ * Defines the connection between the {@link AzureComputeApi} implementation and the jclouds
+ * {@link org.jclouds.compute.ComputeService}.
+ */
+@Singleton
+public class AzureComputeServiceAdapter implements ComputeServiceAdapter<VMDeployment, VMHardware, VMImage, Location> {
+
+   private static int runningNumber = 1;
+   private final String azureGroup;
+
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   private Logger logger = Logger.NULL;
+
+   private final AzureComputeApi api;
+
+   private final AzureComputeConstants azureComputeConstants;
+
+   @Inject
+   AzureComputeServiceAdapter(final AzureComputeApi api, final AzureComputeConstants azureComputeConstants,
+                              @Named("azureGroupId") String group) {
+
+      this.api = api;
+      this.azureComputeConstants = azureComputeConstants;
+      this.azureGroup = group;
+   }
+
+   @Override
+   public NodeAndInitialCredentials<VMDeployment> createNodeWithGroupEncodedIntoName(
+           final String group, final String name, final Template template) {
+
+      final String location = template.getLocation().getId();
+      DeploymentTemplateBuilder deploymentTemplateBuilder = api.deploymentTemplateFactory().create(group, name, template);
+
+      final String loginUser = deploymentTemplateBuilder.getLoginUserUsername();
+      final String loginPassword = deploymentTemplateBuilder.getLoginPassword();
+
+      DeploymentBody deploymentTemplateBody =  deploymentTemplateBuilder.getDeploymentTemplate();
+
+      DeploymentProperties properties = DeploymentProperties.create(deploymentTemplateBody);
+
+      final String deploymentTemplate = UrlEscapers.urlFormParameterEscaper().escape(deploymentTemplateBuilder.getDeploymentTemplateJson(properties));
+
+      logger.debug("Deployment created with name: %s", name);
+
+      final Set<VMDeployment> deployments = Sets.newHashSet();
+
+      ResourceGroupApi resourceGroupApi = api.getResourceGroupApi();
+      ResourceGroup resourceGroup = resourceGroupApi.get(azureGroup);
+      final String resourceGroupName;
+
+      if (resourceGroup == null){
```

Comment:
Race conditions will appear here when deploying N nodes, as jclouds will call this method concurrently. All "shared" resources should be created *before* creating the actual nodes. This can be done by subclassing the `CreateNodesWithGroupEncodedIntoNameThenAddToSet` class and overriding its `execute` method. Look at how [DigitalOcean](https://github.com/jclouds/jclouds/blob/master/providers/digitalocean2/src/main/java/org/jclouds/digitalocean2/compute/strategy/CreateKeyPairsThenCreateNodes.java) creates the key pairs before creating the nodes and how [it binds that strategy](https://github.com/jclouds/jclouds/blob/master/providers/digitalocean2/src/main/java/org/jclouds/digitalocean2/compute/config/DigitalOcean2ComputeServiceContextModule.java#L92) to the Guice context.
Moving this logic there will eliminate the race condition and the code here will be able to assume the resource group always exists.